src/libhirte/log/log.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <systemd/sd-journal.h>
+
+#include "libhirte/common/common.h"
+
+typedef enum LogLevel {
+        LOG_LEVEL_DEBUG,
+        LOG_LEVEL_INFO,
+        LOG_LEVEL_WARN,
+        LOG_LEVEL_ERROR,
+} LogLevel;
+
+const char *log_level_to_string(LogLevel l);
+
+
+typedef void (*LogFn)(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data);
+
+void hirte_log_to_journald_with_location(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data);
+void hirte_log_to_stdout_with_location(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data);
+
+struct HirteLogConfig {
+        LogFn log_fn;
+        LogLevel level;
+        bool is_quiet;
+};
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+extern struct HirteLogConfig config;
+
+void hirte_log_set_log_fn(LogFn log_fn);
+void hirte_log_set_level(LogLevel level);
+void hirte_log_set_quiet(bool is_quiet);
+bool shouldLog(LogLevel lvl);
+
+#define hirte_log(lvl, msg, ...)                                                          \
+        if (shouldLog(lvl)) {                                                             \
+                config.log_fn(lvl, __FILE__, _SD_STRINGIFY(__LINE__), __func__, msg, ""); \
+        }
+
+#define hirte_log_debug(msg) hirte_log(LOG_LEVEL_DEBUG, msg)
+#define hirte_log_info(msg) hirte_log(LOG_LEVEL_INFO, msg)
+#define hirte_log_warn(msg) hirte_log(LOG_LEVEL_WARN, msg)
+#define hirte_log_error(msg) hirte_log(LOG_LEVEL_ERROR, msg)
+
+
+#define hirte_logf(lvl, msgfmt, ...)                                                      \
+        if (shouldLog(lvl)) {                                                             \
```

Comment:
This is slightly weird do me. Either you should make extern struct HirteLogConfig config public in the header like it is now (although then it needs a better name to not globally conflict as "config"), but then "shouldLog should be inline so that everything in hirte_logf() gets inlined. 

Or alternatively, make the config private to log.c and then make the inner part of this if a function call that does on the call to config.log in log.c where it doesn't need to be public.

Personally I would prefer to have the config only visilble in log.c

Reply:
I didn't know how to pass `_VA_ARGS_` to a a function, but using your suggested `vasprintf` and passing the `va_list` works well. Could you please have another look?